dmscripts/export_dos_suppliers.py

```diff
@@ -0,0 +1,275 @@
+import os
+import sys
+from multiprocessing.pool import ThreadPool
+
+from dmapiclient import HTTPError
+from dmscripts.insert_dos_framework_results import (
+    CORRECT_DECLARATION_RESPONSE_MUST_BE_TRUE, CORRECT_DECLARATION_RESPONSE_MUST_BE_FALSE,
+    CORRECT_DECLARATION_RESPONSE_SHOULD_BE_FALSE, MITIGATING_FACTORS,
+    CORRECT_DECLARATION_RESPONSES
+)
+
+if sys.version_info[0] < 3:
+    import unicodecsv as csv
+else:
+    import csv
+
+FRAMEWORK_SLUG = 'digital-outcomes-and-specialists'
+DRAFT_STATUSES = [
+    'completed',
+    'failed',
+    'draft',
+]
+LOTS = [
+    'digital-outcomes', 'digital-specialists',
+    'user-research-studios', 'user-research-participants',
+]
+
+
+def find_suppliers(client, framework_slug):
+    suppliers = client.get_interested_suppliers(framework_slug)['interestedSuppliers']
+    return ({'supplier_id': supplier_id} for supplier_id in suppliers)
+
+
+def add_supplier_info(client):
+    def inner(record):
+        supplier = client.get_supplier(record['supplier_id'])
+
+        return dict(record,
+                    supplier=supplier['suppliers'])
+
+    return inner
+
+
+def add_framework_info(client, framework_slug):
+    def inner(record):
+        try:
+            supplier_framework = client.get_supplier_framework_info(record['supplier']['id'], framework_slug)
+            supplier_framework = supplier_framework['frameworkInterest']
+            supplier_framework['declaration'] = supplier_framework['declaration'] or {}
+
+            if supplier_framework['declaration'].get('status') != 'complete':
+                assert supplier_framework['onFramework'] is False, \
+                    "Incomplete declaration but not failed, have you inserted the framework result?"
+
+            return dict(record,
+                        declaration=supplier_framework['declaration'],
+                        onFramework=supplier_framework['onFramework'])
+        except HTTPError as e:
+            if e.status_code != 404:
```

Comment:
I'm not sure we'd want to fail silently for a 404 here as it shouldn't ever happen as far as I can think? The suppliers returned by `find_suppliers` should by definition be only those suppliers that have a `supplier_frameworks` entry for DOS. So a 404 here should probably set some alarm bells ringing.